native/src/main/scala/sqs4s/auth/Credentials.scala

```diff
@@ -95,73 +92,73 @@ object Credentials {
       secretKey
     ).pure[F].widen[Credential]))
 
-  def envVarResource[F[_]: Sync]: Resource[F, Credentials[F]] =
-    Resource.liftF(envVar[F])
-
-  private def envVar[F[_]: Sync]: F[Credentials[F]] =
-    Sync[F].delay {
-      val ACCESS_KEY_ENV_VAR = "AWS_ACCESS_KEY_ID"
-      val ALTERNATE_ACCESS_KEY_ENV_VAR = "AWS_ACCESS_KEY"
-
-      val SECRET_KEY_ENV_VAR = "AWS_SECRET_KEY"
-      val ALTERNATE_SECRET_KEY_ENV_VAR = "AWS_SECRET_ACCESS_KEY"
-
-      val AWS_SESSION_TOKEN_ENV_VAR = "AWS_SESSION_TOKEN"
-      val optAccessKey = Option(System.getenv(ACCESS_KEY_ENV_VAR)) |+| Option(
-        System.getenv(ALTERNATE_ACCESS_KEY_ENV_VAR)
-      )
-      val optSecretKey = Option(System.getenv(SECRET_KEY_ENV_VAR)) |+| Option(
-        System.getenv(ALTERNATE_SECRET_KEY_ENV_VAR)
-      )
-      val optSessionToken = Option(System.getenv(AWS_SESSION_TOKEN_ENV_VAR))
-
-      instance[F] {
-        (optAccessKey, optSecretKey, optSessionToken) match {
-          case (Some(accessKey), Some(secretKey), None) =>
-            BasicCredential(accessKey, secretKey)
-              .pure[F].widen
-
-          case (Some(accessKey), Some(secretKey), Some(token)) =>
-            TemporarySecurityCredential(
-              accessKey,
-              secretKey,
-              token
-            ).pure[F].widen
-
-          case _ =>
-            Sync[F].raiseError[Credential](NoEnvironmentVariablesFound)
+  def envVar[F[_]: Sync]: Resource[F, Credentials[F]] =
+    Resource.liftF {
+      Sync[F].delay {
+        val ACCESS_KEY_ENV_VAR = "AWS_ACCESS_KEY_ID"
+        val ALTERNATE_ACCESS_KEY_ENV_VAR = "AWS_ACCESS_KEY"
+
+        val SECRET_KEY_ENV_VAR = "AWS_SECRET_KEY"
+        val ALTERNATE_SECRET_KEY_ENV_VAR = "AWS_SECRET_ACCESS_KEY"
+
+        val AWS_SESSION_TOKEN_ENV_VAR = "AWS_SESSION_TOKEN"
+        val optAccessKey = Option(System.getenv(ACCESS_KEY_ENV_VAR)) |+| Option(
+          System.getenv(ALTERNATE_ACCESS_KEY_ENV_VAR)
+        )
+        val optSecretKey = Option(System.getenv(SECRET_KEY_ENV_VAR)) |+| Option(
+          System.getenv(ALTERNATE_SECRET_KEY_ENV_VAR)
+        )
+        val optSessionToken = Option(System.getenv(AWS_SESSION_TOKEN_ENV_VAR))
+
+        instance[F] {
+          (optAccessKey, optSecretKey, optSessionToken) match {
+            case (Some(accessKey), Some(secretKey), None) =>
+              BasicCredential(accessKey, secretKey)
+                .pure[F].widen
+
+            case (Some(accessKey), Some(secretKey), Some(token)) =>
+              TemporarySecurityCredential(
+                accessKey,
+                secretKey,
+                token
+              ).pure[F].widen
+
+            case _ =>
+              Sync[F].raiseError[Credential](NoEnvironmentVariablesFound)
+          }
```

Comment:
So we still have the problem that the checking for the env vars is happening during the credentials access and not creation.
This means that if we use `chain` it will always use this one because its creation will never fail.

We can do this instead:

```scala
Sync[F].defer {
  // ...

  (optAccessKey, optSecretKey, optSessionToken) match {
    case (Some(accessKey), Some(secretKey), None) =>
      Credentials.basic[F](accessKey, secretKey).pure[F]

    case (Some(accessKey), Some(secretKey), Some(token)) =>
      Credentials.temporal[F](accessKey, secretKey, token).pure[F]

    case _ =>
      Sync[F].raiseError[Credential](NoEnvironmentVariablesFound)
  }
}
```

Where `Credentials.temporal` is implemented like `Credentials.basic`.